### PR TITLE
ISPN-1342 Near caching for Java Hot Rod client

### PR DESCRIPTION
**Pull request by galderz**
https://issues.jboss.org/browse/ISPN-1342

* Add near cache functionality to Java Hot Rod client.
* Two modes implemented, lazy and eager.
* Near caches can optionally be configured to keep a maximum number of entries.
* Refactored Hot Rod client testing utilities.
* Upon near cache failover, near caches are cleared.

### CODE REVIEW

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8182/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22536&buildTypeId=bt9&guest=1) for a93e3c0b59cedfa6464047f04ca8651919646240
Tests: 8788, 0 failed (0 new), 6 ignored. Build time: 00:20:36

Error message is logged

Process exited with code 1

Compilation failure

**Comment by galderz** [comment]
Compilation error with recently added test, fixing...

**Comment by galderz** [comment]
Compilation error fixed

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8185/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22545&buildTypeId=bt9&guest=1) for d1a8258b4d9b74eecc130c67b4df492d0ae5d5fb
Tests: 8788, 0 failed (0 new), 6 ignored. Build time: 00:31:39

Error message is logged

Process exited with code 1

Compilation failure

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/near/LinkedMapNearCache.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   protected LinkedMapNearCache(LinkedHashMap<K, VersionedValue<V>> cache) {
+      this.cache = cache;
+   }
+
+   @Override
+   public void put(K key, VersionedValue<V> value) {
+      Lock lock = rwlock.writeLock();
+      try {
+         lock.lock();
+         cache.put(key, value);
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/near/LinkedMapNearCache.java]: According to the LinkedHashMap javadoc, reinsertion does not cause the insertion order to be changed, therefore the eviction handled by removeEldestEntry() will possibly evict a recently refreshed entry. Not really a problem, but a possible optimization for the future. Did you think about initializing the LinkedHashMap to use accessOrder ? Would it degrade performance ?

File: documentation/src/main/asciidoc/user_guide/chapter-48-Using_Hot_Rod_Server.adoc [hunk 2]
```diff
@@ -562,6 +562,58 @@ missed, it could decide to clear any locally cached data when the fail over
...
+...
+
+// For lazy near caches
+ConfigurationBuilder lazy = new ConfigurationBuilder();
+lazy.nearCache().mode(NearCacheMode.LAZY);
+
+// For eager near caches
+ConfigurationBuilder eager = new ConfigurationBuilder();
+eager.nearCache().mode(NearCacheMode.EAGER);
+----
```
Comment (tristantarrant) [inline documentation/src/main/asciidoc/user_guide/chapter-48-Using_Hot_Rod_Server.adoc]: Add an example of optionally setting a maximum number of entries

**Comment by tristantarrant** [comment]
Two minor comments, but this looks excellent.

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8186/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22546&buildTypeId=bt9&guest=1) for d1a8258b4d9b74eecc130c67b4df492d0ae5d5fb
Tests: 11229, 1 failed (1 new), 23 ignored. Build time: 02:48:10

Failed tests detected
* (new) TestSuite: org.infinispan.statetransfer.ClusterTopologyManagerTest.testClusterRecoveryWithRebalance



Error message is logged

Build duration (secs) is too large: 10049

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8199/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22561&buildTypeId=bt9&guest=1) for b44dafa4a099b9ff3c0ec840b00648e754d8aacb
Tests: 11228, 4 failed (4 new), 23 ignored. Build time: 01:40:40

Failed tests detected
* (new) TestSuite: org.infinispan.distribution.DistSyncTxL1RepeatableReadFuncTest.testBackupOwnerInvalidatesL1WhenPrimaryIsUnaware


* (new) TestSuite: org.infinispan.interceptors.ConcurrentInterceptorVisibilityTest.testSizeVisibility


* (new) TestSuite: org.infinispan.client.hotrod.near.EvictEagerNearCacheTest.testEvictAfterReachingMax


* (new) TestSuite: org.infinispan.client.hotrod.near.ClusterEagerNearCacheTest.testNearCacheUpdatesSeenByAllClients

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity Infinispan :: Pull requests monitor [Build 8222/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22599&buildTypeId=bt9&guest=1) for b44dafa4a099b9ff3c0ec840b00648e754d8aacb
Tests: 11230, 0 failed (0 new), 23 ignored. Build time: 01:41:40

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8242/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22650&buildTypeId=bt9&guest=1) for b44dafa4a099b9ff3c0ec840b00648e754d8aacb
Tests: 11231, 2 failed (2 new), 23 ignored. Build time: 02:00:41

Failed tests detected
* (new) TestSuite: org.infinispan.distexec.mapreduce.DistributedFourNodesMapReduceTest.testInvokeMapReduceOnAllKeysWithCollator


* (new) org.infinispan.server.test.query.ManualIndexingIT.testManualIndexing



Error message is logged

Build duration (secs) is too large: 6800

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/near/LinkedMapNearCache.java (hunk 1 above)
Comment (galderz) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/near/LinkedMapNearCache.java]: I think you are right, it should be initialized with `accessOrder=true`, good catch.

File: documentation/src/main/asciidoc/user_guide/chapter-48-Using_Hot_Rod_Server.adoc (hunk 2 above)
Comment (galderz) [inline documentation/src/main/asciidoc/user_guide/chapter-48-Using_Hot_Rod_Server.adoc]: Yeah, forgot that.

**Comment by galderz** [comment]
@tristantarrant Updated PR fixing those two things.

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8251/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22672&buildTypeId=bt9&guest=1) for 335857b7e78ab7426e05dd78d0d9edc2f45e2e45
Tests: 11231, 1 failed (1 new), 23 ignored. Build time: 01:41:20

Failed tests detected
* (new) TestSuite: org.infinispan.interceptors.ConcurrentInterceptorVisibilityTest.testSizeVisibility

**Comment by tristantarrant** [comment]
Nice one Galder, it's in.

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity Infinispan :: Pull requests monitor [Build 8256/galderz:t_1342_x](http://ci.infinispan.org/viewLog.html?buildId=22683&buildTypeId=bt9&guest=1) for 335857b7e78ab7426e05dd78d0d9edc2f45e2e45
Tests: 11230, 1 failed (1 new), 23 ignored. Build time: 02:10:38

Failed tests detected
* (new) TestSuite: org.infinispan.client.hotrod.near.ClusterLazyNearCacheTest.testNearCacheUpdatesSeenByAllClients



Error message is logged

Build duration (secs) is too large: 7822

**Pull request closed without merging**